File: samples/swiftui/Sources/Services/SampleEnvironment.swift
import Foundation
import MilanoSDK

#if canImport(UIKit)
    import UIKit
#elseif canImport(AppKit)
    import AppKit
#endif

/// The sample's Milano setup: one engine, the design system registered,
/// builders per screen. Screens depend on this service, never on engine
/// internals.
final class SampleEnvironment {
    static let shared = SampleEnvironment()

    private let observer = ConsoleObserver()
    private let analytics = ConsoleAnalytics()
    private let engine: MilanoEngine

    private init() {
        do {
            // The engine keeps the contract default: unknown types fail the
            // build. Surfaces that can degrade gracefully opt into skip below.
            engine = try MilanoEngine(
                vocabularyJSON: Self.resource("vocabulary"),
                registry: MilanoBridge.registry(),
                observer: observer,
                userInteractionObserver: analytics)
            SampleVocabulary.assertMatches(engine)
        } catch {
            fatalError("Milano engine setup failed: \(error)")
        }
    }

    /// One shared context for every screen: each document reads only the
    /// keys it declares; the rest are ignored by rule.
    private static let sharedContext: [String: MilanoValue] = [
        "userName": .string("Ada"),
        "marketingConsentRequired": .bool(true)
    ]

    /// Self-contained documents (banners, the tip calculator): context
    /// injected; any declared state gets instant defaults. A screen may add
    /// its own context values on top of the shared ones (the Pokemon demo
    /// injects what it fetched); on a key collision the screen wins.
    func documentBuilder(
        resource: String, screenContext: [String: MilanoValue] = [:]
    ) -> MilanoViewBuilder {
        // Banners are optional, promotional surfaces: an unknown component
        // degrades to a gap instead of failing the build. The form and the
        // interstitial keep the fail default; their content is load-bearing.
        let builder = engine.viewBuilder(document: Self.resource(resource))
        if resource.hasPrefix("banner") {
            builder.unknownTypePolicy(.skip)
        }
        return builder
            .context(Self.sharedContext.merging(screenContext) { _, screen in screen })
            .stateData { declarations in Self.defaults(for: declarations) }
            .actionHandler(Self.handle(_:))
            .label(resource)
    }

    /// The interstitial: the document's `dismiss` action is interpreted by
    /// the presenting screen; every other action takes the shared path.
    func interstitialBuilder(onDismiss: @escaping @Sendable () -> Void) -> MilanoViewBuilder {
        engine.viewBuilder(document: Self.resource("interstitial"))
            .context(Self.sharedContext)
            .actionHandler { action in
                if case .dismiss = SampleAction(action) {
                    await MainActor.run { onDismiss() }
                    return nil
                }
                return try await Self.handle(action)
            }
            .label("interstitial")
    }

    /// The form: initial values arrive through the async state data
    /// provider, as if fetched from an API.
    func formBuilder() -> MilanoViewBuilder {
        engine.viewBuilder(document: Self.resource("contact-form"))
            .context(Self.sharedContext)
            .stateData { declarations in
                try await Task.sleep(nanoseconds: 700_000_000)
                return Self.defaults(for: declarations)
            }
            .actionHandler(Self.handle(_:))
            .label("contact-form")
    }

    // MARK: - Provider defaults

    private static func defaults(for declarations: [String: MilanoType]) -> [String: MilanoValue] {
        declarations.mapValues { type in
            if type.optional { return .null }
            switch type.kind {
            case .bool: return .bool(false)
            case .int: return .int(0)
            case .double: return .double(0)
            default: return .string("")
            }
        }
    }

    // MARK: - Action funnel

    /// The single async funnel: navigation and submission live in the host.
    /// Generated bindings make the switch typed and exhaustive. The returned
    /// value is the completion result: submitContact declares `result:
    /// "string"`, so its confirmation number flows back into the document's
    /// onSuccess actions as the `result` root.
    @Sendable private static func handle(_ action: MilanoAction) async throws -> MilanoValue? {
        switch SampleAction(action) {
        case .openUrl(let urlString):
            if let url = URL(string: urlString) {
                #if canImport(UIKit)
                    await MainActor.run { UIApplication.shared.open(url) }
                #elseif canImport(AppKit)
                    await MainActor.run { NSWorkspace.shared.open(url) }
                #endif
            }
            return nil
        case .submitContact(let name, let surname, let email, let phone):
            // Simulated network call; the returned confirmation number is
            // what a real backend would answer with.
            print("[sample] submitting \(name) \(surname) <\(email)> \(phone ?? "-")")
            try await Task.sleep(nanoseconds: 1_000_000_000)
            return .string("MC-\(UUID().uuidString.prefix(6))")
        case .dismiss:
            // Interpreted by the presenting screen's handler; inert here.
            return nil
        case .unrecognized(let action):
            print("[sample] unhandled action \(action.name)")
            return nil
        }
    }

    private static func resource(_ name: String) -> Data {
        guard let url = Bundle.main.url(forResource: name, withExtension: "json"),
            let data = try? Data(contentsOf: url)
        else {
            fatalError("missing bundled resource \(name).json")
        }
        return data
    }
}

File: samples/compose/app/src/main/kotlin/dev/getmilano/sample/environment/SampleEnvironment.kt
package dev.getmilano.sample.environment

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import dev.getmilano.MilanoAction
import dev.getmilano.MilanoActionHandler
import dev.getmilano.MilanoEngine
import dev.getmilano.MilanoMainDispatcher
import dev.getmilano.MilanoObserver
import dev.getmilano.MilanoType
import dev.getmilano.MilanoUnknownTypePolicy
import dev.getmilano.MilanoUserInteractionObserver
import dev.getmilano.MilanoValue
import dev.getmilano.MilanoViewBuilder
import dev.getmilano.sample.milanobridge.ExamplesAction
import dev.getmilano.sample.milanobridge.ExamplesVocabulary
import dev.getmilano.sample.milanobridge.milanoRegistry
import dev.getmilano.sample.ui.Screen
import dev.getmilano.viewBuilder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext

/**
 * The sample's Milano setup: one engine, the design system registered,
 * builders per screen. Screens depend on this service, never on engine
 * internals.
 */
class SampleEnvironment(
    private val context: Context,
) {
    /** Logs every occurrence the engine reports: the sample's telemetry. */
    private val observer =
        MilanoObserver { occurrence ->
            Log.d("milano", "${occurrence.kind} view=${occurrence.viewIdentity} node=${occurrence.node ?: "-"}")
        }

    /**
     * The sample's analytics sink: a real app would forward each record to
     * its tracker; the sample logs it. Milano implements no tracker.
     */
    private val analytics =
        MilanoUserInteractionObserver { interaction ->
            Log.d(
                "analytics",
                "${interaction.kind} view=${interaction.viewIdentity}" +
                    " node=${interaction.node ?: "-"} name=${interaction.name ?: "-"}",
            )
        }

    /**
     * One shared context for every screen: each document reads only the
     * keys it declares; the rest are ignored by rule.
     */
    private val sharedContext =
        mapOf(
            "userName" to MilanoValue.StringValue("Ada"),
            "marketingConsentRequired" to MilanoValue.BoolValue(true),
        )

    private val engine: MilanoEngine by lazy {
        // The engine keeps the contract default: unknown types fail the
        // build. Surfaces that can degrade gracefully opt into skip below.
        MilanoEngine(
            vocabularyJson = asset("vocabulary.json"),
            registry = milanoRegistry(),
            observer = observer,
            userInteractionObserver = analytics,
        ).also { ExamplesVocabulary.assertMatches(it) }
    }

    /** The single async funnel: navigation and submission live in the host. */
    private val handler = MilanoActionHandler { action -> handle(action) }

    fun builder(screen: Screen): MilanoViewBuilder =
        when (screen) {
            Screen.FORM -> formBuilder()
            else -> documentBuilder(screen.key)
        }

    /**
     * The interstitial: the document's `dismiss` action is interpreted by
     * the presenting screen; every other action takes the shared path.
     */
    fun interstitialBuilder(onDismiss: () -> Unit): MilanoViewBuilder =
        engine
            .viewBuilder(asset("interstitial.json"))
            .context(sharedContext)
            .actionHandler { action ->
                if (ExamplesAction.from(action) is ExamplesAction.Dismiss) {
                    withContext(Dispatchers.Main) { onDismiss() }
                    null
                } else {
                    handle(action)
                }
            }.dispatcher(MilanoMainDispatcher())
            .label("interstitial")

    /**
     * The Pokemon demo: the screen fetches its own values first, then adds
     * them on top of the shared context for this one host.
     */
    fun pokemonBuilder(screenContext: Map<String, MilanoValue>): MilanoViewBuilder = documentBuilder("pokemon", screenContext)

    /**
     * The profile screen: identity values a real app would fetch from its
     * account service, injected as screen context over the shared context.
     */
    fun profileBuilder(): MilanoViewBuilder =
        documentBuilder(
            "profile",
            mapOf(
                "memberSince" to MilanoValue.StringValue("March 2024"),
                "avatarUrl" to
                    MilanoValue.StringValue(
                        "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/25.png",
                    ),
            ),
        )

    /**
     * Self-contained documents (banners, the expression demos): context
     * injected; any declared state gets instant defaults. A screen may add
     * its own context values on top of the shared ones; on a key collision
     * the screen wins.
     */
    private fun documentBuilder(
        resource: String,
        screenContext: Map<String, MilanoValue> = emptyMap(),
    ): MilanoViewBuilder =
        engine
            .viewBuilder(asset("$resource.json"))
            .apply {
                // Banners are optional, promotional surfaces: an unknown
                // component degrades to a gap instead of failing the build.
                // The form and the interstitial keep the fail default.
                if (resource.startsWith("banner")) unknownTypePolicy(MilanoUnknownTypePolicy.SKIP)
            }.context(sharedContext + screenContext)
            .stateDataProvider { declarations -> defaults(declarations) }
            .actionHandler(handler)
            .dispatcher(MilanoMainDispatcher())
            .label(resource)

    /**
     * The form: initial values arrive through the async state data
     * provider, as if fetched from an API.
     */
    private fun formBuilder(): MilanoViewBuilder =
        engine
            .viewBuilder(asset("contact-form.json"))
            .context(sharedContext)
            .stateDataProvider { declarations ->
                delay(700)
                defaults(declarations)
            }.actionHandler(handler)
            .dispatcher(MilanoMainDispatcher())
            .label("contact-form")

    private fun defaults(declarations: Map<String, MilanoType>): Map<String, MilanoValue> =
        declarations.mapValues { (_, type) ->
            when {
                type.optional -> MilanoValue.Null
                type.kind is MilanoType.Kind.Bool -> MilanoValue.BoolValue(false)
                type.kind is MilanoType.Kind.Int -> MilanoValue.IntValue(0)
                type.kind is MilanoType.Kind.Double -> MilanoValue.DoubleValue(0.0)
                else -> MilanoValue.StringValue("")
            }
        }

    /**
     * The returned value is the completion result: submitContact declares
     * result "string", so its confirmation number flows back into the
     * document's onSuccess actions as the result root.
     */
    private suspend fun handle(action: MilanoAction): MilanoValue? {
        // Generated bindings make the dispatch typed and exhaustive.
        when (val decoded = ExamplesAction.from(action)) {
            is ExamplesAction.OpenUrl -> {
                val intent = Intent(Intent.ACTION_VIEW, Uri.parse(decoded.url))
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                context.startActivity(intent)
            }

            is ExamplesAction.SubmitContact -> {
                // Simulated network call; the returned confirmation number
                // is what a real backend would answer with.
                Log.d("sample", "submitting ${decoded.name} ${decoded.surname} <${decoded.email}>")
                delay(1_000)
                return MilanoValue.StringValue("MC-${java.util.UUID.randomUUID().toString().take(6)}")
            }

            is ExamplesAction.Dismiss -> {
                // Interpreted by the presenting screen's handler; inert here.
            }

            is ExamplesAction.Unrecognized -> {
                Log.d("sample", "unhandled action ${decoded.action.name}")
            }
        }
        return null
    }

    private fun asset(name: String): String =
        context.assets
            .open(name)
            .bufferedReader()
            .use { it.readText() }
}
